spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java

```diff
@@ -365,9 +367,16 @@ private void doReleaseMessage(Message<?> message) {
 
 	private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
 		if (this.messageStore instanceof SimpleMessageStore) {
-			SimpleMessageGroup messageGroup =
-					(SimpleMessageGroup) this.messageStore.getMessageGroup(this.messageGroupId);
-			return messageGroup.remove(message);
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				if (messages.contains(message)) {
```

Comment:
This is not so good - messages is an unmodified collection wrapping a `LinkedBlockingQueue`. `contains` has to iterate over the contents which is not good for a large group.

We had a similar problem in the `AcceptOnceFileListFilter` and we solved that (INT-3572) by maintaining a `Set` parallel to the queue; we can't use that technique here, though.

Looking at `SimpleMessageGroup` - it's really not so good for large groups - `remove()` requires iterating over the entire queue as well.

This probably explains Ryan's performance observation.

We should look at using a (synchronized) `LinkedHashSet` instead of the `BlockingQueue` and compare the performance for both small and large groups.

If using a synchronized set is too expensive, we might have to end up with 2 stores - one for small groups and one for large.